### PR TITLE
vimPlugins.leetcode-nvim: init at 2024-06-27

### PR DESCRIPTION
**Pull request by nvimtor**
## Description of changes
Adds the plugin [leetcode.nvim](https://github.com/kawre/leetcode.nvim).

Includes hard dependencies `telescope-nvim` and `nui-nvim`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by linsui** [comment]
Close https://github.com/NixOS/nixpkgs/issues/274061

**Review by PerchunPak — APPROVED** [review]
LGTM! The order doesn't matter as it is broken anyway

Fixes #274061

CC @GaetanLepage

**Review by GaetanLepage — APPROVED** [review]

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 1]
```diff
@@ -2198,6 +2198,10 @@
     dependencies = with self; [ plenary-nvim ];
   };
 
+  leetcode-nvim = super.leetcode-nvim.overrideAttrs {
+    dependencies = with self; [ nui-nvim telescope-nvim ];
```
Comment (GaetanLepage) [inline pkgs/applications/editors/vim/plugins/overrides.nix:2202]: Could you add
```
doInstallCheck = true;
nvimRequireCheck = "leetcode";
```
please ?

Comment (nvimtor) [reply pkgs/applications/editors/vim/plugins/overrides.nix:2202]: Will do!

**Comment by nvimtor** [comment]
Also updated the SHA as it was incorrect; apologies. I double checked it with `nurl` and it should be the correct hash now.

**Review by GaetanLepage — APPROVED** [review]
I have made minor cosmetic changes.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 340378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Looks like dependencies are not transitive: `telescope-nvim` has `plenary-nvim` as a dependency, but `leetcode-nvim` won't see it (despite it having `telescope-nvim` as a dependency).

cc @teto (I have been pinging you a lot lately, sorry for that)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 340378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 340378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
```
Check whether the following module can be imported: leetcode
Error detected while processing pre-vimrc command line:
E518: Unknown option: /nix/store/9711k560h24xp38drsrdzjawfm3v6i2p-vimplugin-lua5.1-telescope.nvim-scm-1-unstable-2024-09-11
Entering Ex mode.  Type "visual" to go to Normal mode.
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
We need to fix the `require` check.

**Comment by PerchunPak** [comment]
Blocked by #342240 ([tracker](nixpk.gs/pr-tracker.html?pr=342240))

**Comment by PerchunPak** [comment]
> ha indeed. The neovimRequireCheckHook just adds the listed dependencies to RTP. a quickfix could be to add `vimUtils.transitiveClosure dependencies` to rtp instead. It still wont work for plugins that pull python provider/extra programs dependencies etc but could work for 90% plugins. Ideally we would use the same utilities to generate the wrapped neovim and the tests. We are getting there :)

https://github.com/NixOS/nixpkgs/pull/342240 fixes it, no?

**Comment by teto** [comment]
> https://github.com/NixOS/nixpkgs/pull/342240 fixes it, no?

I dont think so, it will just expand the derivation `dependencies` without the transitives ones, as mentioned by gaetan.

**Comment by PerchunPak** [comment]
`git cherry-pick b8581496dfb668d839e879c8e17393d44b726887 dc2ec23437620454af9f0beb4b96e5eb2a21d816 && nix-build -A vimPlugins.leetcode-nvim` passes, so... it is fixed by https://github.com/NixOS/nixpkgs/pull/342240

**Comment by GaetanLepage** [comment]
> git cherry-pick [b858149](https://github.com/NixOS/nixpkgs/commit/b8581496dfb668d839e879c8e17393d44b726887) [dc2ec23](https://github.com/NixOS/nixpkgs/commit/dc2ec23437620454af9f0beb4b96e5eb2a21d816) && nix-build -A vimPlugins.leetcode-nvim builds succesful, so... it is fixed by #342240

@nvimtor can you go ahead and rebase this then please ?

**Comment by PerchunPak** [comment]
> nvimtor can you go ahead and rebase this then please ?

No, because b8581496dfb668d839e879c8e17393d44b726887 is https://github.com/NixOS/nixpkgs/pull/342240, which is currently in staging-next and not in master (https://nixpk.gs/pr-tracker.html?pr=342240)

**Comment by GaetanLepage** [comment]
> No, because [b858149](https://github.com/NixOS/nixpkgs/commit/b8581496dfb668d839e879c8e17393d44b726887) is #342240, which is currently in staging-next and not in master (https://nixpk.gs/pr-tracker.html?pr=342240)

Oh right, I got tricked by the "merged" status.

**Comment by PerchunPak** [comment]
b8581496dfb668d839e879c8e17393d44b726887 got to master, so this can be rebased and merged :tada:

**Comment by GaetanLepage** [comment]
Formatting issue.

**Comment by nvimtor** [comment]
@PerchunPak @GaetanLepage done; just to clarify, do we still need `plenary.nvim` to be listed a dependency?

**Comment by nvimtor** [comment]
Ah, let me run nix fmt.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 340378`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>vimPlugins.leetcode-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
> do we still need `plenary.nvim` to be listed as a dependency?

Can you try without it ?
The `requireCheck` should catch a failure if it would happen to not propagate from telescope.

**Comment by nvimtor** [comment]
@GaetanLepage Great, it does fail without `plenary.nvim`: 

```
error: builder for '/nix/store/xl6s5s6ahy5ih6bw1cpnnznplrh31scc-vimplugin-leetcode.nvim-2024-06-27.drv' failed with exit code 1;
       last 30 log lines:
       > Executing vimCommandCheckHook
       > Running phase: neovimRequireCheckHook
       > Executing neovimRequireCheckHook
       > Check whether the following module can be imported: leetcode
       > Error detected while processing pre-vimrc command line:
       > E5108: Error executing lua ...in-leetcode.nvim-2024-06-27/lua/leetcode/config/init.lua:2: module 'plenary.path' not found:
       >  no field package.preload['plenary.path']
       >       no file './plenary/path.lua'
       >   no file '/nix/store/m8nihfxqggb2zgvvd7nd7rmd8rb3y52h-luajit-2.1.1713773202/share/luajit-2.1/plenary/path.lua'
       >  no file '/usr/local/share/lua/5.1/plenary/path.lua'
       >    no file '/usr/local/share/lua/5.1/plenary/path/init.lua'
       >       no file '/nix/store/m8nihfxqggb2zgvvd7nd7rmd8rb3y52h-luajit-2.1.1713773202/share/lua/5.1/plenary/path.lua'
       >     no file '/nix/store/m8nihfxqggb2zgvvd7nd7rmd8rb3y52h-luajit-2.1.1713773202/share/lua/5.1/plenary/path/init.lua'
       >        no file './plenary/path.so'
       >    no file '/usr/local/lib/lua/5.1/plenary/path.so'
       >       no file '/nix/store/m8nihfxqggb2zgvvd7nd7rmd8rb3y52h-luajit-2.1.1713773202/lib/lua/5.1/plenary/path.so'
       >        no file '/usr/local/lib/lua/5.1/loadall.so'
       >    no file './plenary.so'
       >         no file '/usr/local/lib/lua/5.1/plenary.so'
       >    no file '/nix/store/m8nihfxqggb2zgvvd7nd7rmd8rb3y52h-luajit-2.1.1713773202/lib/lua/5.1/plenary.so'
       >     no file '/usr/local/lib/lua/5.1/loadall.so'
       > stack traceback:
       >  [C]: in function 'require'
       >     ...in-leetcode.nvim-2024-06-27/lua/leetcode/config/init.lua:2: in main chunk
       >   [C]: in function 'require'
       >     ...ffvi-vimplugin-leetcode.nvim-2024-06-27/lua/leetcode.lua:1: in main chunk
       >   [C]: in function 'require'
       >     [string ":lua"]:1: in main chunk
       > Entering Ex mode.  Type "visual" to go to Normal mode.
       > :
       For full logs, run 'nix log /nix/store/xl6s5s6ahy5ih6bw1cpnnznplrh31scc-vimplugin-leetcode.nvim-2024-06-27.drv'.
error: 1 dependencies of derivation '/nix/store/6jir5jw9w1j6xnzf8sjhnkg5zliblabs-neovim-0.10.2.drv' failed to build
```

**Comment by GaetanLepage** [comment]
> it does fail without `plenary.nvim`:

Ok, so it looks like transitive dependencies are indeed broken. cc @teto

**Comment by GaetanLepage** [comment]
Merging as is then

**Review by GaetanLepage — APPROVED** [review]

**Pull request merged**